packages/field-base/src/helper-controller.js

```diff
@@ -0,0 +1,190 @@
+/**
+ * @license
+ * Copyright (c) 2021 Vaadin Ltd.
+ * This program is available under Apache License Version 2.0, available at https://vaadin.com/license/
+ */
+import { SlotController } from '@vaadin/component-base/src/slot-controller.js';
+
+/**
+ * A controller manage the helper node content.
+ */
+export class HelperController extends SlotController {
+  constructor(host) {
+    // Do not provide slot factory, as only create helper lazily.
+    super(host, 'helper');
+  }
+
+  get helperId() {
+    return this.node && this.node.id;
+  }
+
+  /** @protected */
+  get uniqueId() {
+    return SlotController.helperId;
+  }
+
+  /**
+   * Override to initialize the newly added custom helper.
+   *
+   * @param {Node} helperNode
+   * @protected
+   * @override
+   */
+  initCustomNode(helperNode) {
+    this.__updateHelperId(helperNode);
+
+    this.__observeHelper(helperNode);
+
+    const hasHelper = this.__hasHelper(helperNode);
+    this.__toggleHasHelper(hasHelper);
+  }
+
+  /**
+   * Override to cleanup helper node when it's removed.
+   *
+   * @param {Node} _node
+   * @protected
+   * @override
+   */
+  teardownNode(_node) {
+    // The observer does not exist when the default helper is removed.
+    if (this.__helperIdObserver) {
+      this.__helperIdObserver.disconnect();
+    }
+
+    this.__applyDefaultHelper(this.helperText);
+  }
+
+  /**
+   * Set helper text based on corresponding host property.
+   * @param {string} helperText
+   */
+  setHelperText(helperText) {
+    this.helperText = helperText;
+
+    this.__applyDefaultHelper(helperText);
+  }
+
+  /**
+   * @param {HTMLElement} helperNode
+   * @private
+   */
+  __applyCustomHelper(helperNode) {
+    this.__updateHelperId(helperNode);
+
+    const hasHelper = this.__hasHelper(helperNode);
+    this.__toggleHasHelper(hasHelper);
+  }
```

Comment:
It seems that this method is not used anywhere.

Reply:
Removed.